src/java.desktop/share/classes/javax/swing/SwingWorker.java

```diff
@@ -301,18 +301,16 @@ public SwingWorker() {
                 new Callable<T>() {
                     public T call() throws Exception {
                         setState(StateValue.STARTED);
-                        return doInBackground();
+                        try {
+                            return doInBackground();
+                        } finally {
+                            setState(StateValue.DONE);
+                            doneEDT();
+                        }
```

Comment:
Probably been discussed already, but this will not follow this part of spec:
```
 * Before the {@code doInBackground} method is invoked on a <i>worker</i> thread,
 * {@code SwingWorker} notifies any {@code PropertyChangeListeners} about the
 * {@code state} property change to {@code StateValue.STARTED}.  After the
 * {@code doInBackground} method is finished the {@code done} method is
 * executed.  Then {@code SwingWorker} notifies any {@code PropertyChangeListeners}
 * about the {@code state} property change to {@code StateValue.DONE}.
```

Reply:
Valid concern.

I mentioned it before [here](https://github.com/openjdk/jdk/pull/11940#issuecomment-1407690445), yet I forgot about this later on.

Changing the order of the calls should be enough.

I guess we should verify this condition in the test, what do you think?

Reply:
But it did not work as expected before the fix as well since the "notification and done" work in parallel anyway.

Reply:
It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".

Reply:
I haven't verified it, but I think it worked as it's stated: `done` is called first, then listeners are notified. According to the order of calls

https://github.com/openjdk/jdk/blob/b00b70c2400d28070d26630614a010bc52237827/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L310-L313

`SwingWorker.done` gets invoked first followed by `setState` which, in its turn, notifies the listeners.

At this time, as you pointed out, the order is reversed:

https://github.com/openjdk/jdk/blob/76b02ea3a8add954705d6df7f91f5896d141b02d/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L307-L308

Reply:
> It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".

Is flushing required?

Two `AtomicBoolean`s should do the job. EDT can't process two events simultaneously, one follows the other. So, `done` is expected to be called while both are false, and it sets its own flag to `true`; the listener should be called second, so its flag is false and `done`'s flag is true.

I'll leave it to Pransanta to figure out the details.

Reply:
It seems the order of sequence is
listener->State.STARTED->doInBackground->listener->DONE->done

Reply:
> It seems the order of sequence is listener->State.STARTED->doInBackground->listener->DONE->done

Yes, but the different order is specified: listener(STARTED) -> doInBackground -> done -> listener(DONE).

Reply:
But then it will violate
https://github.com/openjdk/jdk/blob/810c8a271b4524ae776e2306ef699e04a7d145a2/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

Reply:
Hm… not really, state will transition to `DONE` after `doInBackground` is finished.

In fact, it'll happen soon enough: `doneEDT` will schedule execution of `done` on EDT and exit. Then the state will change to `DONE`. Yet listeners are notified on EDT, to this notification will also be scheduled to run on EDT.

Very much likely that the state is `DONE` when `done` is executed but it's not guaranteed.

---

The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to `DONE` as soon as `doInBackground` completes. Then both `done` method and listeners are called when the state is `DONE` that is *the background work is complete*.

Yet as it's specified now, the `done` method is called while the state is still `STARTED`. This kind of implies calling `get` may block because the result of computation may not be available yet. However, it's not the case: `get` never blocks if called from `done`: it either returns the result or throws `CancellationException`.

Reply:
> The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to DONE as soon as doInBackground completes. Then both done method and listeners are called when the state is DONE that is the background work is complete.

Which parts are contradicts? I mean that the old implementation call DONE-listener last, as specified in the header of the class what part of the spec mention different behavior? Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?

Reply:
> > The specification isn't clear, and some parts somewhat contradict each other. Logically, the state should transition to `DONE` as soon as `doInBackground` completes. Then both done method and listeners are called when the state is `DONE` that is the background work is complete.
> 
> Which parts are contradicts?

This was a wrong statement. In fact, I referred to the problem that is being fixed here: `done` may be called before `doInBackground`.

https://github.com/openjdk/jdk/blob/0aeebee284effe9abd0ed3cf2845430b40bb53bd/src/java.desktop/share/classes/javax/swing/SwingWorker.java#L288-L293

The spec for `DONE` state does not specify when the transition occurs.

> I mean that the old implementation call DONE-listener last, as specified in the header of the class what part of the spec mention different behavior?

Yes, this is how it's specified now.

> Above we discussed that it is possible to see a difference if the listener will do some work on EDT, after the fix the listener will not be called last, is that safe to assume it does not break something?

What kind of difference? Yes, you can observe the fact that now the `DONE`-state-listener is called before the `done` method.

In either case, the listener and the method are called when `get` does not block. The background job is finished when either of these events occurs.

I see no problem here.

Reply:
> > It could be tested by flushing and blocking the EDT in the listener. That will allow us to check the order of events posted from the listener and from the "doneEDT".
> 
> Is flushing required?
> 
> Two `AtomicBoolean`s should do the job.

I referred to `isStateDoneNotified` and `isDoneCalled`.

> EDT can't process two events simultaneously, one follows the other. So, `done` is expected to be called while both are `false`, and it sets its own flag to `true`; the listener should be called second, so its flag is `false` and `done`'s flag is `true`.

Do we want to ensure it?